projects/chaos/src/relay_fsm.c

```diff
@@ -0,0 +1,154 @@
+}
+
+static StatusCode prv_ack_open_callback(CANMessageID msg_id, uint16_t device, CANAckStatus status,
+                                        uint16_t num_remaining, void *context) {
+  RelayFsmAckCtx *ack_ctx = context;
+  if (status != CAN_ACK_STATUS_OK) {
+    if (ack_ctx->retries < RELAY_FSM_MAX_RETRIES) {
+      ack_ctx->retries++;
+      event_raise(CHAOS_EVENT_RETRY_RELAY, ack_ctx->id);
+      return STATUS_CODE_OK;
+    } else {
+      ack_ctx->retries = 0;
+      event_raise(CHAOS_EVENT_RELAY_ERROR, ack_ctx->id);
+      return STATUS_CODE_OK;
+    }
+  }
+
+  ack_ctx->retries = 0;
+  event_raise(CHAOS_EVENT_RELAY_OPENED, ack_ctx->id);
+  return STATUS_CODE_OK;
+}
+
+static StatusCode prv_ack_close_callback(CANMessageID msg_id, uint16_t device, CANAckStatus status,
+                                         uint16_t num_remaining, void *context) {
+  RelayFsmAckCtx *ack_ctx = context;
+  if (status != CAN_ACK_STATUS_OK) {
+    if (ack_ctx->retries < RELAY_FSM_MAX_RETRIES) {
+      ack_ctx->retries++;
+      event_raise(CHAOS_EVENT_RETRY_RELAY, ack_ctx->id);
+      return STATUS_CODE_OK;
+    } else {
+      ack_ctx->retries = 0;
+      event_raise(CHAOS_EVENT_RELAY_ERROR, ack_ctx->id);
+      return STATUS_CODE_OK;
+    }
+  }
+
+  ack_ctx->retries = 0;
+  event_raise(CHAOS_EVENT_RELAY_CLOSED, ack_ctx->id);
+  return STATUS_CODE_OK;
+}
+
+FSM_DECLARE_STATE(relay_opened);
+
+FSM_DECLARE_STATE(closing_relay);
+
+FSM_DECLARE_STATE(relay_closed);
+
+FSM_DECLARE_STATE(opening_relay);
+
+FSM_STATE_TRANSITION(relay_opened) {
+  FSM_ADD_GUARDED_TRANSITION(CHAOS_EVENT_CLOSE_RELAY, prv_guard, closing_relay);
+}
+
+FSM_STATE_TRANSITION(closing_relay) {
+  FSM_ADD_GUARDED_TRANSITION(CHAOS_EVENT_RELAY_CLOSED, prv_guard, relay_closed);
+  FSM_ADD_GUARDED_TRANSITION(CHAOS_EVENT_RETRY_RELAY, prv_guard, closing_relay);
+}
+
+FSM_STATE_TRANSITION(relay_closed) {
+  FSM_ADD_GUARDED_TRANSITION(CHAOS_EVENT_OPEN_RELAY, prv_guard, opening_relay);
+}
+
+FSM_STATE_TRANSITION(opening_relay) {
+  FSM_ADD_GUARDED_TRANSITION(CHAOS_EVENT_RELAY_OPENED, prv_guard, relay_opened);
+  FSM_ADD_GUARDED_TRANSITION(CHAOS_EVENT_RETRY_RELAY, prv_guard, opening_relay);
+}
+
+static void prv_relay_transmit(RelayId id, RelayState state, const CANAckRequest *ack_request) {
+  switch (id) {
+    case RELAY_ID_SOLAR_MASTER_FRONT:
+      // TODO(ELEC-105): Replace with solar front/back.
+      CAN_TRANSMIT_SOLAR_RELAY(ack_request, state);
+      return;
+    case RELAY_ID_SOLAR_MASTER_REAR:
+      CAN_TRANSMIT_SOLAR_RELAY(ack_request, state);
+      return;
+    case RELAY_ID_BATTERY:
+      CAN_TRANSMIT_BATTERY_RELAY(ack_request, state);
+      return;
+    case RELAY_ID_MAIN_POWER:
+      CAN_TRANSMIT_MAIN_RELAY(ack_request, state);
+      return;
+    case NUM_RELAY_IDS:
+    default:
+      return;
+  }
+}
+
+static void prv_opening(FSM *fsm, const Event *e, void *context) {
+  RelayFsmCtx *relay_ctx = context;
+  relay_ctx->request.callback = prv_ack_open_callback;
```

Comment:
You could probably use the same function for open/close and just use the event? It might not be as obvious how it works though. This approach is more explicit at the expense of duplicated code.

Reply:
I think given how complex the rest of this FSM is we should at least have separate state transition functions.